platform/linux-generic/odp_event_validation.c

```diff
@@ -0,0 +1,232 @@
+#include <odp_libconfig_internal.h>
+#include <odp_macros_internal.h>
+#include <odp_print_internal.h>
+
+#include <inttypes.h>
+#include <string.h>
+
+#define EVENT_VALIDATION_NONE  0
+#define EVENT_VALIDATION_WARN  1
+#define EVENT_VALIDATION_ABORT 2
+
+#define EVENT_DATA_PRINT_MAX_LEN 128
+
+typedef struct {
+	odp_atomic_u64_t err_count[_ODP_EV_MAX];
+	odp_shm_t shm;
+
+} event_validation_global_t;
+
+typedef struct {
+	const char *str;
+} _odp_ev_info_t;
+
+static event_validation_global_t *_odp_ev_glb;
+
+#if _ODP_EVENT_VALIDATION
+
+/* Table for mapping function IDs to API function names */
+static const _odp_ev_info_t ev_info_tbl[] = {
+	[_ODP_EV_BUFFER_FREE]       = {.str = "odp_buffer_free()"},
+	[_ODP_EV_BUFFER_FREE_MULTI] = {.str = "odp_buffer_free_multi()"},
+	[_ODP_EV_BUFFER_IS_VALID]   = {.str = "odp_buffer_is_valid()"},
+	[_ODP_EV_EVENT_FREE]        = {.str = "odp_event_free()"},
+	[_ODP_EV_EVENT_FREE_MULTI]  = {.str = "odp_event_free_multi()"},
+	[_ODP_EV_EVENT_FREE_SP]     = {.str = "odp_event_free()_sp"},
+	[_ODP_EV_EVENT_IS_VALID]    = {.str = "odp_event_is_valid()"},
+	[_ODP_EV_PACKET_FREE]       = {.str = "odp_packet_free()"},
+	[_ODP_EV_PACKET_FREE_MULTI] = {.str = "odp_packet_free_multi()"},
+	[_ODP_EV_PACKET_FREE_SP]    = {.str = "odp_packet_free_sp()"},
+	[_ODP_EV_PACKET_IS_VALID]   = {.str = "odp_packet_is_valid()"}
+};
+
+ODP_STATIC_ASSERT(_ODP_ARRAY_SIZE(ev_info_tbl) == _ODP_EV_MAX, "ev_info_tbl missing entries");
+
+static void print_event_data(odp_event_t event, odp_event_type_t type)
+{
+	const char *type_str;
+	const uint32_t bytes_per_row = 16;
+	uint32_t byte_len;
+	int num_rows, max_len, n;
+	int len = 0;
+	uint8_t *data;
+
+	if (type == ODP_EVENT_PACKET) {
+		odp_packet_t pkt = odp_packet_from_event(event);
+
+		data = odp_packet_data(pkt);
+		byte_len = odp_packet_seg_len(pkt);
+		type_str = "Packet";
+	} else {
+		odp_buffer_t buf = odp_buffer_from_event(event);
+
+		data = odp_buffer_addr(buf);
+		byte_len = odp_buffer_size(buf);
+		type_str = "Buffer";
+	}
+
+	if (byte_len > EVENT_DATA_PRINT_MAX_LEN)
+		byte_len = EVENT_DATA_PRINT_MAX_LEN;
+
+	num_rows = (byte_len + bytes_per_row - 1) / bytes_per_row;
+	max_len = 256 + (3 * byte_len) + (3 * num_rows);
+	n = max_len - 1;
+
+	char str[max_len];
+
+	len += _odp_snprint(&str[len], n - len, "%s %p data %p:\n", type_str, event, data);
+	while (byte_len) {
+		uint32_t row_len = byte_len > bytes_per_row ? bytes_per_row : byte_len;
+
+		len += _odp_snprint(&str[len], n - len, " ");
+
+		for (uint32_t i = 0; i < row_len; i++)
+			len += _odp_snprint(&str[len], n - len, " %02x", data[i]);
+
+		len += _odp_snprint(&str[len], n - len, "\n");
+
+		byte_len -= row_len;
+		data += row_len;
+	}
+
+	_ODP_PRINT("%s\n", str);
+}
+
+static inline int validate_event_endmark(odp_event_t event, _odp_ev_id_t id, odp_event_type_t type)
+{
+	uint64_t *endmark_ptr = _odp_event_endmark_get_ptr(event);
+
+	if (odp_unlikely(*endmark_ptr != _ODP_EV_ENDMARK_VAL)) {
```

Comment:
This could be inverted -> have a likely branch return 0 and then the unlikely stuff outside the if.

Reply:
Changed in V4.